libraries/botbuilder-dialogs-adaptive/src/adaptiveDialogBot.ts

```diff
@@ -0,0 +1,136 @@
+        context.turnState.set(BotFrameworkClientKey, botFrameworkClient);
+        context.turnState.set(skillConversationIdFactoryKey, this.skillConversationIdFactoryBase);
+        context.turnState.set('ConversationState', this.conversationState);
+        context.turnState.set('UserState', this.userState);
+        context.turnState.set(resourceExplorerKey, this.resourceExplorer);
+        context.turnState.set('memoryScopes', this.memoryScopes);
+        context.turnState.set('pathResolvers', this.pathResolvers);
+
+        const languageGenerator = this.resourceExplorer.getResource(this.languageGeneratorId)
+            ? new ResourceMultiLanguageGenerator(this.languageGeneratorId)
+            : new TemplateEngineLanguageGenerator();
+        context.turnState.set(languageGeneratorKey, languageGenerator);
+
+        let manager: LanguageGeneratorManager;
+        if (!AdaptiveDialogBot.languageGeneratorManagers.has(this.resourceExplorer)) {
+            manager = new LanguageGeneratorManager(this.resourceExplorer);
+            AdaptiveDialogBot.languageGeneratorManagers.set(this.resourceExplorer, manager);
+        } else {
+            manager = AdaptiveDialogBot.languageGeneratorManagers.get(this.resourceExplorer);
+        }
+        context.turnState.set(languageGeneratorManagerKey, manager);
+
+        context.turnState.set(languagePolicyKey, this.languagePolicy);
+        context.turnState.set(BotTelemetryClientKey, this.telemetryClient);
+
+        // Catch "setTestOptions" event and save into "conversation.testOptions".
+        // Note: This is consumed by AdaptiveExpressions Extensions.RandomNext
+        if (context.activity.type === ActivityTypes.Event && context.activity.name === 'setTestOptions') {
+            console.log('setTestOptions received. This could change the behavior of AdaptiveExpressions RandomNext');
+
+            await context.turnState
+                .get<ConversationState>('ConversationState')
+                .createProperty('testOptions')
+                .set(context, context.activity.value);
+        }
+
+        // Put this on the TurnState using set because some adapters (like BotFrameworkAdapter and CloudAdapter) will have already added it.
+        context.turnState.set(BotCallbackHandlerKey, this.onTurn);
+    }
+
+    private createDialog(): Dialog {
+        const adaptiveDialogResource = this.resourceExplorer.getResource(this.adaptiveDialogId);
+        if (!adaptiveDialogResource) {
+            throw new Error(`The ResourceExplorer could not find a resource with id "${this.adaptiveDialogId}"`);
+        }
+
+        const adaptiveDialog = this.resourceExplorer.loadType<AdaptiveDialog>(adaptiveDialogResource);
+
+        // If we were passed any Dialogs then add them to the AdaptiveDialog's DialogSet so they can be invoked from any other Dialog.
+        this.dialogs.forEach((dialog) => {
+            adaptiveDialog.dialogs.add(dialog);
+        });
```

Comment:
I've heard enough about `Dialogs` being available to others and nested `DialogSets` to believe that this should have a unit test to verify and lock-in the behavior we expect. @joshgummersall or @mdrichardson can you add a unit test for this `forEach`? Thanks!

Reply:
Added!